include/numerics/dense_matrix_base_impl.h

```diff
@@ -23,14 +23,23 @@
 // Local Includes
 #include "libmesh/dense_matrix.h"
 #include "libmesh/dense_vector_base.h"
+#include "libmesh/dense_vector.h"
 
 // C++ includes
 #include <iomanip> // for std::setw()
 
 namespace libMesh
 {
 
-
+  template <typename T>
+  DenseVector<T>
+  DenseMatrixBase<T>::diagonal() const
+  {
+    DenseVector<T> ret;
+    for (decltype(_m) i = 0; i < _m; ++i)
+      ret(i) = el(i, i);
+    return std::move(ret);
```

Comment:
There's an Effective C++11 item about returning by `std::move`. Basically, you aren't supposed to do it, because it can inhibit RVO. Instead just `return ret;`.

Reply:
https://stackoverflow.com/questions/4986673/c11-rvalues-and-move-semantics-confusion-return-statement

Reply:
Haha - We've had a number of debates on the "decltype" question. It boils down to whether you trust contributors to use the right index type. In my experience the answer is no. Depending on the background and experience level of your contributor, you'll get either signed or unsigned variants of ints. That might result in a sign mismatch warning if you turn on enough compiler warning or worse you might actually overflow the int type if you are working with really big ranges.

In MOOSE we created a "mooseindex(type)" call that will always give you the right indexing type back for the type you pass in. It works for containers as well so you don't have to do even ugly tricks to get a non-const POD type back in those cases. Still many in our community (and our team) still dislike it...